feat: cross-chain swaps ui re-design

micaelae opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
Changes include
- layout and styling updates for the landing screen
- re-styling input fields, including asset pickers
- integrates the multichain asset list into bridge's token list generator
- style + copy updates in quote display components
- advanced settings modal
- input and quote validation alerts
- bug fixes

This depends on 3 open PRs
1. tracking events: https://github.com/MetaMask/metamask-extension/pull/28713
2. multichain AssetPicker: https://github.com/MetaMask/metamask-extension/pull/28975
3. NetworkAvatar style update: https://github.com/MetaMask/metamask-extension/pull/28976

Changes from #2 and #3 are currently included here, but will mainly just contain bridge component updates after those are merged. Since those are being reviewed by external teams, reviews on this one should be focused on bridge-specific files/directories

Figma: https://www.figma.com/design/IuOIRmU3wI0IdJIfol0ESu/Cross-Chain-Swaps?node-id=7-24563&node-type=canvas&m=dev

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/28373?quickstart=1)

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/MMS-1451

## **Manual testing steps**

1. Set these in .metamaskrc and run `yarn webpack --watch`
```
SEGMENT_HOST='http://localhost:9090'
SEGMENT_WRITE_KEY='FAKE'
BRIDGE_USE_DEV_APIS=1
```
2. Try out the Bridge page, asset picker, submitting txs, viewing quotes etc
3. Open the background console network tab to see emitted events

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->
![Screenshot 2024-12-11 at 3 02 20 PM](https://github.com/user-attachments/assets/826f7cba-202d-4a52-b15f-ca16c29d7a96)


### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [X] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

micaelae committed:
fix: set chain before navigating to bridge experience

micaelae committed:
chore: bridge transaction settings modal + slippage

micaelae committed:
chore: style bridge tooltip

micaelae committed:
chore: alternative-soft color definitions

micaelae committed:
chore: use multichain asset type for src/dest token lists

micaelae committed:
chore: useBridgeTokens hook that loads address allowlists

micaelae committed:
chore: style quote card

micaelae committed:
chore: fix quotes modal colors

micaelae committed:
fix: trim whitespace from formatted token amounts

micaelae committed:
chore: validate bridge-api token list response

micaelae committed:
chore: bridge asset-picker button

micaelae committed:
fix: navigate to bridge when HW wallet is connected

micaelae committed:
chore: add size props to MM Mascot animation

micaelae committed:
chore: style CTA button

micaelae committed:
chore: update page layout

micaelae committed:
chore: update countdown timer hook

micaelae committed:
chore: banner alerts for validations

micaelae committed:
chore: style input fields

micaelae committed:
fix: incorrect quote sorting

micaelae committed:
chore: add loading state for bridge tokens

micaelae committed:
test: wip prepare-bridge-page tests

infiniteflower reviewed (commented):
Things that need a pointer on hover:
1. "More quotes"
2. Best price question mark tooltip

Also I think I broke something by clicking the switch From/To, the animation was so mesmerizing, I just kept clicking it. Started with Arbitrum ETH > Ethereum ETH, clicked switch a bunch of times, and wound up with Ethereum ETH > Ethereum ETH.

https://github.com/user-attachments/assets/37fb6bb4-63ed-46f5-92e5-20cd5861b480

micaelae committed:
fix: rm bridge input field style

micaelae committed:
chore: display multiple network avatars in PickerNetwork + AvatarGroup

micaelae committed:
fix: bridge slippage modal style + input

micaelae committed:
fix: reset dest token if dest chain is selected as src chain

micaelae committed:
fix: unset selected toToken on dest network change

micaelae committed:
chore: shorten bridge network names in quote card

micaelae committed:
fix: bridge asset picker button

micaelae committed:
chore: hardware wallet messaging

micaelae committed:
feat: shrink input text as char count increases

micaelae committed:
chore: shorten token amounts

micaelae committed:
chore: show shorter bridge network names

micaelae committed:
fix: set timeout between switch button clicks

micaelae committed:
fix: tmp network icon

micaelae committed:
fix: snapshot tests

micaelae committed:
chore: expire bridge quotes after 30s

micaelae committed:
chore: enable unimported networks as destination

micaelae committed:
chore: prevent wrapping in quote card

micaelae committed:
chore: indicate when ETA is <1 min

micaelae committed:
fix: hide gas balance alert when user has insufficient src amount balance

micaelae committed:
fix: blank CTA button when refreshing unsubmittable quote

micaelae committed:
fix: update snapshots

micaelae committed:
fix: unit tests

micaelae committed:
chore: fetch exchange rates for selected token with unimported network

micaelae committed:
fix: hide 0 token balance

micaelae committed:
fix: only track network change if new value is different

micaelae committed:
fix: include value in Input Changed events

micaelae committed:
chore: misc style updates

micaelae committed:
fix: snapshots

micaelae committed:
fix: selectors unit tests

micaelae committed:
fix: type issues

micaelae committed:
fix: double borders

micaelae committed:
fix: scss lint issues

micaelae committed:
fix: snapshots

micaelae committed:
fix: unit tests

micaelae committed:
fix: useLatestBalance type

micaelae committed:
chore: click to copy dest address

micaelae committed:
chore: refactor token list and add tests

micaelae committed:
test: update snapshots

micaelae committed:
use multichain asset in bridge

georgewrmarshall commented on ui/helpers/constants/design-system.ts:
```diff
@@ -54,6 +54,7 @@ export enum Color {
 export enum BackgroundColor {
   backgroundDefault = 'background-default',
   backgroundAlternative = 'background-alternative',
+  backgroundAlternativeSoft = 'background-alternative-soft',
```
We just released a version of the design-tokens `@metmask/design-tokens@v4.2.0` with updated background colors `background-muted`, `background-muted-hover` and `background-muted-active`. Should this be 

```suggestion
  backgroundMuted = 'background-muted',
  backgroundMutedHover = 'background-muted-hover',
  backgroundMutedActive = 'background-muted-active',
```

You will also need to update the CSS I believe. I'll create a PR

georgewrmarshall commented on ui/helpers/constants/design-system.ts:
```diff
@@ -109,6 +110,7 @@ export enum BorderColor {
 export enum TextColor {
   textDefault = 'text-default',
   textAlternative = 'text-alternative',
+  textAlternativeSoft = 'text-alternative-soft',
```
Not aware of this color does `textMuted` not work?

micaelae replied:
It's a new color introduced in [Cross-chain Swaps](https://www.figma.com/design/IuOIRmU3wI0IdJIfol0ESu/Cross-Chain-Swaps?node-id=1697-15289&node-type=instance&m=dev). It's not exactly the same as textMuted

Is it ok to do this for now and update the design package as a fast follow? I just have the color defined in css

georgewrmarshall replied:
It might be a disparity between the versions of colors in figma and the extension we text-alternative (soft) is just `text-alternative` I'll ask to remove the brackets to prevent confusion. This PR should resolve any color value discrepancies https://github.com/MetaMask/metamask-extension/pull/29117

georgewrmarshall commented on ui/helpers/constants/design-system.ts:
```diff
@@ -139,6 +141,7 @@ export enum TextColor {
 export enum IconColor {
   iconDefault = 'icon-default',
   iconAlternative = 'icon-alternative',
+  iconAlternativeSoft = 'icon-alternative-soft',
```
Same here. The design-tokens may just need to be updated

georgewrmarshall reviewed (commented):
Hey @micaelae, I think this PR needs to be updated with the latest design tokens version and it should cover you for color. I'll create a PR to add the new background muted

micaelae committed:
fix: unused translations

micaelae committed:
fix: test data

micaelae committed:
fix: prepare-bridge-page tests

micaelae committed:
fix: unit tests

micaelae committed:
fix: show total cost

micaelae committed:
chore: add autofocus prop to AssetPicker

micaelae committed:
fix: e2e tests

micaelae committed:
chore: rm TODO

infiniteflower commented on ui/ducks/bridge/selectors.ts:
```diff
@@ -243,11 +274,31 @@ export const getToTokenConversionRate = createDeepEqualSelector(
   (state) => ({
     state,
     toTokenExchangeRate: state.bridge.toTokenExchangeRate,
+    toTokenUsdExchangeRate: state.bridge.toTokenUsdExchangeRate,
```
Not a blocker, but these could be selectors

infiniteflower commented on ui/ducks/bridge/selectors.ts:
```diff
@@ -515,7 +524,7 @@ export const getValidationErrors = createDeepEqualSelector(
       // Shown after fetching quotes
       isInsufficientGasForQuote: (balance?: BigNumber) => {
-        if (balance && activeQuote && fromToken) {
+        if (balance && activeQuote && fromToken && fromTokenInputValue) {
```
So we only want to calculate insufficient gas for quote only after the user has typed something in.

infiniteflower commented on ui/pages/bridge/bridge.util.ts:
```diff
@@ -124,7 +125,12 @@ export async function fetchBridgeTokens(
+        TOKEN_VALIDATORS.concat(TOKEN_AGGREGATOR_VALIDATORS),
+        token,
+        url,
+        false, // Don't log errors for tokens
```
Ah yea I think I was seeing a ton of errors for tokens in the logs, thanks for this.

infiniteflower commented on ui/pages/bridge/prepare/bridge-cta-button.tsx:
```diff
@@ -36,18 +48,40 @@ export const BridgeCTAButton = () => {
 
-  const { isNoQuotesAvailable, isInsufficientBalance } =
-    useSelector(getValidationErrors);
+  const {
+    isNoQuotesAvailable,
+    isInsufficientBalance: isInsufficientBalance_,
+    isInsufficientGasBalance: isInsufficientGasBalance_,
+    isInsufficientGasForQuote: isInsufficientGasForQuote_,
+  } = useSelector(getValidationErrors);
```
I just noticed that some of these are functions, usually selectors just return data. I think these probably should go in a hook or something. Not a blocker though.

infiniteflower reviewed (dismissed):
I'm fine with approving this if we address my small nits in a follow up PR some other time. Great work!

jclancy93 commented:
> Doing some testing, my observations:
> 
>     Best price tool tip and More Quotes buttons are missing hover pointers
>     Can't see tokens I own in Bridge token list unless they are imported first (but I think that's intentional? or an artifact of the other token importing design)

I noticed both of these as well

micaelae committed:
fix: tests

Gudahtt commented on ui/pages/bridge/utils/quote.ts:
```diff
@@ -224,7 +236,7 @@ export const formatCurrencyAmount = (
   if (precision === 0) {
     if (amount.lt(0.01)) {
-      return `<${formatCurrency('0', currency, precision)}`;
+      return '<$0.01';
```
Not a blocker, as this is a pre-existing issue, but this is not localized correctly. We should be using `Intl.NumberFormat` for all number/currency displays, passing in the currency symbol and locale so that it's rendered correctly.

Some locales use `,` instead of `.`, and of course currency symbols can differ (maybe this is always USD? 🤔  But still)

micaelae replied:
will note this as a fast-follow task!

Gudahtt commented on ui/pages/bridge/utils/validators.ts:
```diff
@@ -53,6 +54,15 @@ export const FEATURE_FLAG_VALIDATORS = [
+    property: 'aggregators',
+    type: 'object',
+    validator: (v: unknown): v is number[] =>
+      isValidObject(v) && Object.values(v).every(isValidString),
```
Is this meant to be `isValidNumber`? Or is the return type wrong here? It's saying that it's validating these to be numbers, but it's not.

Gudahtt replied:
Oh the return type is wrong either way. `v` is an object, not an array of numbers.

Gudahtt commented on ui/pages/bridge/bridge.util.ts:
```diff
@@ -124,7 +125,12 @@ export async function fetchBridgeTokens(
     if (
-      validateResponse<SwapsTokenObject>(TOKEN_VALIDATORS, token, url) &&
+      validateResponse<SwapsTokenObject>(
+        TOKEN_VALIDATORS.concat(TOKEN_AGGREGATOR_VALIDATORS),
```
This is called in the background. I wonder if this is causing the increase in memory usage

micaelae committed:
fix: use single token validator

infiniteflower committed:
chore: bridge, show max gas (#29162)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/29162?quickstart=1)

This PR displays the max gas for a bridge quote.

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

![Screenshot 2024-12-12 at 1 58
59 PM](https://github.com/user-attachments/assets/f087b232-6603-4466-a2c3-e9f23cb62e91)
![Screenshot 2024-12-12 at 1 59
03 PM](https://github.com/user-attachments/assets/10792d34-4f73-4442-b775-305011f3f2bb)


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Jack Clancy <jack.clancy93@gmail.com>

micaelae committed:
fix: rm type import from asset-picker

micaelae committed:
fix: silence token validation errors
